### PR TITLE
fix bugs

### PR DESCRIPTION
**Pull request by rakkit**

### CODE REVIEW

File: torchtitan/components/optimizer.py [hunk 1]
```diff
@@ -507,13 +533,23 @@ def build_optimizers(
...
+    ]:
         optim_implementation = job_config.optimizer.implementation
         assert optim_implementation in ["fused", "foreach", "for-loop"]
 
         fused = optim_implementation == "fused"
         foreach = optim_implementation == "foreach"
 
+        mesh_dim_names = extra_kwargs["world_mesh"].mesh_dim_names
+        ep_enable = "dp_shard_1" in mesh_dim_names or "dp_shard_2" in mesh_dim_names
+        if ep_enable:
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:551]: adamw(fused) dose not work EP for now

File: torchtitan/models/MoEllama/GroupedExperts.py [hunk 2]
```diff
@@ -35,61 +34,73 @@ def __init__(
...
     def setup_ep(self, ep_mesh=None, ep_size=None):
         self.ep_mesh = ep_mesh
         self.ep_size = ep_size if ep_size is not None else 1
-        if self.ep_mesh is not None:
-            self.local_rank = self.ep_mesh.get_local_rank()
-        if ep_size is not None:
-            self.expert_per_rank = self.num_experts // ep_size
-        else:
-            self.expert_per_rank = self.num_experts
+        if ep_size is None or ep_size == 1 or ep_mesh is None:
```
Comment (rakkit) [inline torchtitan/models/MoEllama/GroupedExperts.py:58]: If we do EP, each expert will only keep the local experts. And no longer have redundant copy of others

File: torchtitan/models/MoEllama/GroupedExperts.py [hunk 3]
```diff
@@ -35,61 +34,73 @@ def __init__(
...
-        h = h * torch.bmm(x, self.down_proj[lee:les])
-        out = torch.bmm(h, self.up_proj[lee:les])
-
-        out = ep_comm.combine_tokens(
-            out,
-            self.ep_size,
-            self.ep_mesh,
-            all_tokens_per_expert,
-        )
+        h = self.act_fn(torch.bmm(x, self.gate_proj))
```
Comment (rakkit) [inline torchtitan/models/MoEllama/GroupedExperts.py:101]: Because of EP, we don't need to slice partners via rank.  this gives 3x speed up

File: torchtitan/models/MoEllama/moellama.py [hunk 4]
```diff
@@ -240,8 +260,38 @@ def update(self):
                 else:
                     self.bias.data.add_(adjustment.detach())
 
-                self._accumulated_adjustment *= 0
-                self._num_accumulated = 0
+                self._accumulated_adjustment.zero_()
+                self._num_accumulated.zero_()
+
+
+class SharedExperts(nn.Module):
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:267]: for shard experts, let's just assume we have no shard experts. or 1 shared expert.

if we wanna have 2 or more. better to use Linear For NOW.

Because we don't want to add this in to a special case of EP-MUON (3d weights)

File: torchtitan/train.py [hunk 5]
```diff
@@ -472,7 +464,7 @@ def train_step(self, data_iterator: Iterable):
             loss, aux_loss, moe_entropy_per_layer = self.batch_backward(inputs, labels)
             self.metrics_processor.accumulated_losses.append(loss.detach())
             if aux_loss is not None:
-                self.metrics_processor.accumulated_aux_losses.append(loss.detach())
+                self.metrics_processor.accumulated_aux_losses.append(aux_loss.detach())
```
Comment (rakkit) [inline torchtitan/train.py:467]: bug fix

File: torchtitan/train.py [hunk 6]
```diff
@@ -483,21 +475,14 @@ def train_step(self, data_iterator: Iterable):
...
-            but the only 1/ep_size of the parameters are updated.
-            """
-            for model in model_parts:
-                for p_name, p in model.named_parameters():
-                    if "feed_forward.experts" in p_name:
-                        if p.grad is not None:
-                            p.grad = p.grad * parallel_dims.ep
-                        else:
-                            raise Exception(f"p_name: {p_name} and p.grad is None")
+        # # TODO(JSC): if we shard the MoE model, we need to remove the following code
```
Comment (rakkit) [inline torchtitan/train.py:478]: as we don't set grad to 0 manually.

we check if we have a NaN gradient and make sure everything is good.

we can remove this when finalize the code,

File: torchtitan/train.py [hunk 7]
```diff
@@ -483,21 +475,14 @@ def train_step(self, data_iterator: Iterable):
...
+        for model in model_parts:
+            for p_name, p in model.named_parameters():
+                if p.grad is None:
+                    raise Exception(f"p_name: {p_name} and p.grad is None")
 
         grad_norm = None
+        # TODO(JSC): disable gradient clipping for now for debugging
```
Comment (rakkit) [inline torchtitan/train.py:485]: grad clip does not work with EP now

**Comment by janEbert** [comment]
Merged in 02df9feff9829a83ebbbf383491e0fd6082b4d38.
